### DISCUSSION TITLE
Test Ideas for Datastores & Distributed File Systems

### DISCUSSION THREAD

**Discussion by TheENGR** [CS-6381/FinalProject#2] [Ideas]
Hey all, lets come up with some basic ideas for tests!

**Comment by jdunn-git** [comment]
Each of these can be used to evaluate the data store performance for small, medium, and large sets of data - we can decide the values later on:
1. Timing data for writes 
2. Timing data for in order reads
3. Timing data for random order reads
4. Timing data for in order deletes
5. Timing data for random order deletes

**Reply by rich-java-dev** [reply]
We can do this for varying set-size (uniform data contents, but large number of records)
and also for varying pay-loads (consistent number of records, but data-content/size changes)

See other comment for example.

**Comment by rich-java-dev** [comment]
https://cloud.google.com/blog/products/databases/performance-tuning-best-practices-for-memorystore-for-redis
Perhaps we can take some inspiration from redis-benchmark.

**Throughput per data size:**

for example, we want to simulate storing a record of shape:

``` 
{
  id:  random ( or incremented) int/long (or other type if applicable such as string), 
  payLoad: N-sized char/byte string.
}
```

And vary from N from some small size, like Bytes, up through some larger size like 100KB - 1MB.

And measure the AVG time to Read (reads per second), and average time to write (writes per second) over either a pre-defined data-set, or randomly generated one of the given shape.

**Reply by rich-java-dev** [reply]
I think would want some uniform way to to collect the delta times for READS, and for WRITES.

Maybe we can define together a Decorator function for tests done with Python clients that grabs a before/after time, and write to file for analysis.
That way it could be used applied relatively uniformly across the Read/Write functions the Client API exposes, or if there is a better way to get that meta-data from the Data Store server itself that would be good.

![image](https://user-images.githubusercontent.com/32989822/112734333-7719ca80-8f1b-11eb-93ec-c007454ca1dc.png)

**Reply by rich-java-dev** [reply]
I think we would want to hone in on determining sample size/range:
example:
- 1,000 (or 10,000) iterations/data-points (delta times) collected for:
   - Reads
  - Writes
- for each payload of size:
   - 100B
   - 1KB
   - 10KB
   - 100KB

**Reply by rich-java-dev** [reply]
Here is the idea expanded into actual usage:

![image](https://user-images.githubusercontent.com/32989822/112771357-6c356780-8ff9-11eb-8310-9a5d76c1d9d1.png)

**Comment by ekelly4** [comment]
In my research of Ceph, I think that two potential areas for experimentation for the datastores are: 
- consistency (CAP Theorem)
- recovery throughput, rate of healing after a failure
